Fix `declaration-block-no-redundant-longhand-properties` autofix for `transition`

mattxwang opened the pull request:
<!-- Each pull request must be associated with an open issue unless it's a documentation fix. If a corresponding issue does not exist, please create one so we can discuss the change first. -->

<!-- Please answer the following. We close pull requests that don't. -->

> Which issue, if any, is this issue related to?

Closes #6812.

> Is there anything in the PR that needs further explanation?

This one required some non-trivial CSS spec reading, hence the complicated-ish behaviour!

ybiquitous reviewed (changes requested):
@mattxwang Thanks for the pull request! Please consider addressing my reviews (mainly refactoring).

mattxwang commented on lib/rules/declaration-block-no-redundant-longhand-properties/index.js:
```diff
@@ -47,6 +47,39 @@ const customResolvers = new Map([
+				properties.length > 1 &&
+				('none' in properties || 'inherit' in properties || 'initial' in properties)
+			)
+				return;
```
Ah! Great timing, I had just realized this independently of you submitting the review (got sidetracked while writing this up). I added a test case for the unpermitted keyword (this is a dupe of a previous change I made for this autofix), as well as one for `none`.

mattxwang commented:
So sorry for the rushed PR, I got sidetracked (in-person) in the middle of finishing it up - meant to mark it as draft 😅 . Hopefully it wasn't too frustrating to review!

I think I've addressed all of the comments (and slightly changed my approach re: the `none` special case, after re-reading the spec). Let me know what you think!

mattxwang commented on lib/rules/declaration-block-no-redundant-longhand-properties/index.js:
```diff
@@ -47,6 +47,43 @@ const customResolvers = new Map([
+			const timingFunctions = commaSeparated(decls.get('transition-timing-function')?.value);
+			const properties = commaSeparated(decls.get('transition-property')?.value);
+
+			if (!(delays.length && durations.length && timingFunctions.length && properties.length)) {
```
I'm unsure of how to write a test here that isn't invalid syntax (one of the properties would have to be empty) - any thoughts?

ybiquitous replied:
How about specifying an empty declaration value (`transition-delay`) like this?

```css
a { transition-delay: ; transition-duration: 0.1s; transition-timing-function: ease; transition-property: top; }
```

mattxwang replied:
Thanks for the suggestion! Hm, this seems like invalid CSS to me (though I could be wrong) - would that still make it a good test case?

ybiquitous replied:
Surely that CSS seems invalid, but PostCSS can parse it as having an empty value. See the example below:

```js
require("postcss").parse("a { color: ; }")
  .walkDecls(({ prop, value }) => console.log({ prop, value }));
//=> { prop: 'color', value: '' }
```

Adding such a case seems to make sense.

mattxwang replied:
Gotcha; I'm still a bit confused how to proceed, since that case wouldn't be fixable (or in other words, what would I put in `fixed`), but in that case it seems to break the test infrastructure. 

Alternatively, I could extract this out as a utility and have more rigorous tests for the custom resolvers themselves - what do you think?

ybiquitous replied:
How about specifying `unfixable: true` like this?

```js
{
	code: 'a { transition-delay: ; transition-duration: 1s; transition-timing-function: ease; transition-property: top; }',
	unfixable: true,
	message: messages.expected('transition'),
},
```

ybiquitous reviewed (approved):
Thank you! LGTM 👍🏼
